### PR TITLE
Fetch edge by edge id

### PR DESCRIPTION
**Pull request by gvolfing**
Memgraph currently provides id based lookup for vertices, but lacks to ability to do the same for edge ids.
This PR is resolving that issue.

Resolves https://github.com/memgraph/memgraph/issues/1772

### CODE REVIEW

File: src/query/plan/operator.cpp [hunk 1]
```diff
@@ -800,6 +798,38 @@ UniqueCursorPtr ScanAllById::MakeCursor(utils::MemoryResource *mem) const {
...
+  auto edges = [this](Frame &frame, ExecutionContext &context) -> std::optional<std::vector<EdgeAccessor>> {
+    auto *db = context.db_accessor;
+    ExpressionEvaluator evaluator(&frame, context.symbol_table, context.evaluation_context, context.db_accessor, view_);
+    auto value = expression_->Accept(evaluator);
+    if (!value.IsNumeric()) return std::nullopt;
+    int64_t id = value.IsInt() ? value.ValueInt() : value.ValueDouble();
+    if (value.IsDouble() && id != value.ValueDouble()) return std::nullopt;
+    auto maybe_edge = db->FindEdge(storage::Gid::FromInt(id), view_);
+    if (!maybe_edge) return std::nullopt;
+    return std::vector<EdgeAccessor>{*maybe_edge};
```
Comment (antoniofilipovic) [inline src/query/plan/operator.cpp:821]: What is the reason for returning `std::vector` here?

File: src/storage/v2/inmemory/storage.cpp [hunk 2]
```diff
@@ -1429,6 +1456,62 @@ EdgesIterable InMemoryStorage::InMemoryAccessor::Edges(EdgeTypeId edge_type, Vie
...
+    return std::nullopt;
+  }
+
+  auto *edge_ptr = &(*edge_it);
+  auto vertices_acc = mem_storage->vertices_.access();
+
+  auto extract_edge_info = [&](Vertex *from_vertex) -> EdgeInfo {
+    for (auto &out_edge : from_vertex->out_edges) {
+      if (std::get<2>(out_edge).ptr == edge_ptr) {
+        return std::make_tuple(std::get<2>(out_edge), std::get<0>(out_edge), from_vertex, std::get<1>(out_edge));
```
Comment (antoniofilipovic) [inline src/storage/v2/inmemory/storage.cpp:1475]: You can use I think just std::tuple{...} because CTAD will kick in.

**Review by antoniofilipovic — CHANGES_REQUESTED** [review]
Just a few comments. This ended up as a nice solution :muscle:

File: tests/unit/storage_v2_durability_inmemory.cpp [hunk 3]
```diff
@@ -3045,3 +3046,53 @@ TEST_P(DurabilityTest, EdgeTypeIndexRecovered) {
...
+  memgraph::replication::ReplicationState repl_state{memgraph::storage::ReplicationStateRootPath(config)};
+  memgraph::dbms::Database db{config, repl_state};
+  VerifyDataset(db.storage(), DatasetType::ONLY_BASE, GetParam());
+
+  // Check if data has been loaded correctly.
+  {
+    auto acc = db.Access();
+
+    for (auto i{0U}; i < kNumBaseEdges; ++i) {
+      auto edge = acc->FindEdge(memgraph::storage::Gid::FromUint(0), memgraph::storage::View::OLD);
```
Comment (Darych) [inline tests/unit/storage_v2_durability_inmemory.cpp:3082]: What is the purpose on iterating in loop without using `i` variable?

File: tests/unit/storage_v2_durability_inmemory.cpp [hunk 4]
```diff
@@ -3045,3 +3046,53 @@ TEST_P(DurabilityTest, EdgeTypeIndexRecovered) {
...
+    }
+
+    auto edge = acc->FindEdge(memgraph::storage::Gid::FromUint(kNumBaseEdges), memgraph::storage::View::OLD);
+    ASSERT_FALSE(edge.has_value());
+
+    auto vertex = acc->CreateVertex();
+    auto new_edge = acc->CreateEdge(&vertex, &vertex, db.storage()->NameToEdgeType("et"));
+    ASSERT_TRUE(new_edge.HasValue());
+
+    edge = acc->FindEdge(memgraph::storage::Gid::FromUint(kNumBaseEdges + 1), memgraph::storage::View::OLD);
```
Comment (Darych) [inline tests/unit/storage_v2_durability_inmemory.cpp:3093]: Should we check also `auto edge = acc->FindEdge(memgraph::storage::Gid::FromUint(kNumBaseEdges), memgraph::storage::View::OLD);` for true?

File: src/storage/v2/inmemory/storage.cpp [hunk 5]
```diff
@@ -1429,6 +1456,62 @@ EdgesIterable InMemoryStorage::InMemoryAccessor::Edges(EdgeTypeId edge_type, Vie
   return EdgesIterable(mem_edge_type_index->Edges(edge_type, view, storage_, &transaction_));
 }
 
+std::optional<EdgeAccessor> InMemoryStorage::InMemoryAccessor::FindEdge(Gid gid, View view) {
+  using EdgeInfo = std::optional<std::tuple<EdgeRef, EdgeTypeId, Vertex *, Vertex *>>;
```
[the remark this answers is not in the document]
Comment (gvolfing) [reply src/storage/v2/inmemory/storage.cpp:1460]: I would do that, if this part of the code would infect other parts, but this is not the case. I think it's fine to use a tuple in this case and not having to define a separate struct with the members just to have it used as a return type once. Please let me know if you want me to change it.

File: src/storage/v2/edge_accessor.cpp [hunk 6]
```diff
@@ -26,6 +27,15 @@
 #include "utils/memory_tracker.hpp"
 
 namespace memgraph::storage {
+std::optional<EdgeAccessor> EdgeAccessor::Create(EdgeRef edge, EdgeTypeId edge_type, Vertex *from_vertex,
+                                                 Vertex *to_vertex, Storage *storage, Transaction *transaction,
+                                                 View view, bool for_deleted) {
+  if (!IsEdgeVisible(edge.ptr, transaction, view)) {
```
[the remark this answers is not in the document]
Comment (gvolfing) [reply src/storage/v2/edge_accessor.cpp:33]: If a concurrent transaction tries to delete it. We will have to traverse the delta-chain of the Edge object to see if it is valid.

File: src/storage/v2/inmemory/storage.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Darych) [reply src/storage/v2/inmemory/storage.cpp:1460]: No, makes sense. Let's keep it as it is.

File: src/query/plan/operator.cpp [hunk 7]
```diff
@@ -800,6 +798,38 @@ UniqueCursorPtr ScanAllById::MakeCursor(utils::MemoryResource *mem) const {
...
+UniqueCursorPtr ScanAllByEdgeId::MakeCursor(utils::MemoryResource *mem) const {
+  memgraph::metrics::IncrementCounter(memgraph::metrics::ScanAllByEdgeIdOperator);
+
+  auto edges = [this](Frame &frame, ExecutionContext &context) -> std::optional<std::vector<EdgeAccessor>> {
+    auto *db = context.db_accessor;
+    ExpressionEvaluator evaluator(&frame, context.symbol_table, context.evaluation_context, context.db_accessor, view_);
+    auto value = expression_->Accept(evaluator);
+    if (!value.IsNumeric()) return std::nullopt;
+    int64_t id = value.IsInt() ? value.ValueInt() : value.ValueDouble();
+    if (value.IsDouble() && id != value.ValueDouble()) return std::nullopt;
```
[the remark this answers is not in the document]
Comment (Darych) [reply src/query/plan/operator.cpp:818]: I think the original idea of double was to increase range of supported integer numbers :)

**Review by Darych — APPROVED** [review]
LGTM

**Comment by kgolubic** [comment]
@gvolfing can you please provide me with Release notes and related docs PR?

**Pull request merged**